src/bin/pg_autoctl/monitor.c

```diff
@@ -1886,6 +1928,13 @@ monitor_listen_loop(Monitor *monitor)
 			log_warn("Re-establishing connection. We might miss notifications.");
 			pgsql_finish(&(monitor->pgsql));
 
+			if (!ensure_monitor_pg_running(monitor, mconfig))
```

Comment:
This is very weird. Ensure Postgres is running should be done before we try to get notifications, and each time we want to connect to LISTEN to them, just before we connect. 

Maybe having a new function `monitor_listen` that ensure Postgres is running and then calls `pgsql_listen` would be a better design. We would need to make it so that we can call that function first in each loop iteration, so it needs to check if we are already connected and listening ; maybe with a new boolean to track that in the Monitor structure?

Reply:
I made `ensure_monitor_pg_running` to be called before pgsql_listen calls. Also,  terminated active connection monitor->pgsql->connection if the monitor is found to be not running.